File: src/backtrack/No10.py
class No10:
    def isMatch(self, s: str, p: str) -> bool:
        arr = [[0 for i in range(len(p))] for j in range(len(s))]
        
        for i in range(len(p)):
            if p[i] == '*':
                for j in range(i, len(s)):
                    arr[j][i] = 1
            
            else:
                if i == 0 and (s[i] == p[i] or p[i] == '.'):
                    arr[i][i] = 1
                else:
                    for j in range(i-1, len(s)-1):
                        if arr[j][i-1] == 1 and (s[i] == p[i] or p[i] == '.'):
                            arr[j+1][i] = 1
                            break
            
        return arr[len(s)-1][len(p)-1]

File: src/backtrack/No10.java
package src.backtrack;

public class No10 {
    String s;
    String p;
    boolean[][] mem;
    int[][] visit;
    public boolean isMatch(String s, String p) {
        this.s = s;
        this.p = p;
        mem = new boolean[s.length()][p.length()];
        visit = new int[s.length()][p.length()];
        return backtrack(0, 0);
    }

    public boolean backtrack(int i, int j) {
        if (i >= s.length() && j >= p.length()) {
            return true;
        } else if (j >= p.length()) {
            return false;
        } else if (i >= s.length()) {
            while (j < p.length()) {
                if (j + 1 >= p.length()) return false;
                if (p.charAt(j+1) != '*') return false;
                j += 2;
            }
            return true;
        }

        if (visit[i][j] == 1) {
            return mem[i][j];
        }

        visit[i][j] = 1;
        boolean flag = false;
        if (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.') {
            if (j + 1 < p.length() && p.charAt(j+1) == '*') {
                flag = backtrack(i+1, j) || backtrack(i+1, j+2) || backtrack(i, j+2);
            } else {
                flag = backtrack(i+1, j+1);
            }
        } else if (j + 1 < p.length() && p.charAt(j+1) == '*') {
            flag = backtrack(i, j+2);
        }
        mem[i][j] = flag;
        return flag;
    }

    public static void main(String[] args) {
        String s = "bbbba";
        String p = ".*a*a";
        new No10().isMatch(s, p);
    }
}
